File: LABS/LAB6/Solution.java
class Solution {
    public boolean canFinish(int n, int[][] prerequisites) {
        List<Integer>[] adj = new List[n]; //adjacency list for each course
        int[] indegree = new int[n]; //indegrees for each respective course numbered 0..n-1 
        List<Integer> result = new ArrayList<>(); //topological sort's result using Kahn's algorithm

        for (int[] edge : prerequisites) { //edge=[a,b]-->b is a prereq for a so 
          // b needs to be taken before a and therefore, add directed edge from b to a
            int course = edge[0]; 
            int prerequisite = edge[1];
            if (adj[prerequisite] == null) { //no adjacency list in array of list for prereq yet
                adj[prerequisite] = new ArrayList<>(); //create one for course to be added to prereq adj list
            }
            adj[prerequisite].add(course);
            indegree[course]++; //increment course indegree, it has one more prereq pointing to it
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) { //add all courses with prereq/indegree==0 
              //they can be taken without prereq!
              //they're first in topological sort
                queue.offer(i);
            }
        }

        while (!queue.isEmpty()) {
            int current = queue.poll(); //get value at front of queue (lowest indegree/prereq first)
            result.add(current); //add the removed course to the topological sort result
            if (adj[current] != null) { // all adjacent/next courses to be taken 
                for (int next : adj[current]) {
                    indegree[next]--; //have one less prereq (course was removed)
                    if (indegree[next] == 0) {
                        queue.offer(next); //if they now no longer have prereqs, they're next in line
                    }
                }
            }
        }

        return result.size() == n //cycle detection
    }
}

File: LABS/LAB6/solution.py
from collections import deque

def canFinish(numCourses, prerequisites):
    # Create adjacency list for each course
    adj = [[] for _ in range(numCourses)]
    # Indegree for each respective course numbered 0..numCourses-1
    indegree = [0] * numCourses
    # Topological sort result using Kahn's algorithm
    result = []

    # Construct adjacency list and calculate indegrees
    for course, prerequisite in prerequisites:
        adj[prerequisite].append(course)
        indegree[course] += 1

    # Initialize queue with courses having no prerequisites
    queue = deque()
    for i in range(numCourses):
        if indegree[i] == 0:
            queue.append(i)

    # Perform topological sort
    while queue:
        current = queue.popleft()
        result.append(current)
        for next_course in adj[current]:
            indegree[next_course] -= 1
            if indegree[next_course] == 0:
                queue.append(next_course)

    # If all courses are visited, return True, else False
    return len(result) == numCourses

